New command for stand-alone GKE certificates controller

pipejakob opened the pull request:
New stand-alone certificates controller for GKE. Rather than requiring the CA's private key on disk, this allows making external calls to GKE in order to sign cluster certificates.
    
**Which issue this PR fixes**: fixes #39761

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
New GKE certificates controller.
```

CC @mikedanese @jcbsmpsn

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41160)
<!-- Reviewable:end -->

pipejakob commented:
@k8s-bot cvm gce e2e test this
@k8s-bot cri node e2e test this
@k8s-bot kops aws e2e test this

mikedanese commented:
/approve

pipejakob commented:
Hey @lavalamp -- can you approve my hack/verify-flags/known-flags.txt change? Thanks!

mikedanese commented:
@liggitt

liggitt commented on pkg/controller/certificates/webhook_signer_config.go:
```diff
@@ -0,0 +1,63 @@
+	useDefault          = time.Duration(0) // missing/zero value indicates to use default
+)
+
+type webhookSignerConfig struct {
```
config files need versioning. rather than a new config file format, I'd suggest taking a kubeconfig file and a backoff flag set to a reasonable default

pipejakob replied:
Sounds like a reasonable alternative. In my defense, I was following the example of the ImagePolicyWebhook, which uses an unversioned config file:

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/imagepolicy/admission.go#L179

liggitt replied:
Right, which now has tech debt to switch to versioned config :)

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/imagepolicy/admission.go#L217

pipejakob replied:
D'oh. My biggest weakness: TODO comments mere lines away from where I stopped reading :-P.

I like the idea of removing the need for yet another config file anyway, especially since I doubt many users will care enough to override the `retryBackoff`. Will do.

liggitt commented:
it seems weird to have a controller call a webhook... why wouldn't we make the existing signing controller individually runnable and you could run it from wherever you felt safe keeping the private signing key?

mikedanese commented:
Reasons were discussed here https://github.com/kubernetes/kubernetes/issues/39761#issuecomment-272009019

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/41160#issuecomment-278753517
> it seems weird to have a controller call a webhook

https://github.com/kubernetes/kubernetes/issues/39761#issuecomment-272015457
> If we wanted to make a controller that can call out to a webhook, that seems ok, if a little weird

heh, glad to know I agree with myself. I didn't expect webhook capabilities to be added to the controller-manager

mikedanese commented:
FWIW we were planning on creating a webhook approver as well for 1.7. I agree that the controller-manager is bloated but breaking out the certificate controller seems orthogonal to this feature and should be tackled separately. Do you think these webhooks are generally useful features? What is your concerns with adding to the "bloat"? Do you think adding these features will make it significantly harder to break out the certificates controller later on?

liggitt commented:
I think adding webhook call-outs from the controller manager crosses the line.

The point of the API is to allow coordination by external controllers. A signer that wants to closely hold their signing cert/key makes sense to me, but I'd expect that to be handled by the signer running a signing controller themselves, pointed at the kube API:
```
signing-controller --kubeconfig=... --signing-cert=... --signing-key=...
```

If there is no network visibility from the CA to the kube API, and/or the CA wants to expose an API for signing certs, a signing controller that reads from the kube API, calls out to the CA, and updates the kube API with the results could work as well:
```
signing-controller --kubeconfig=... --ca-kubeconfig=...
```

But if network visibility is a concern, it is equally likely the kubernetes installation would not have visibility to the CA, so lumping a glue controller (k8s <-> controller <-> ca) into the controller manager monolith seems like a bad idea.

smarterclayton commented:
The whole design of kube APIs is predicated that multiple controllers can coexist on fundamental resources.  I don't know what @bgrant0607's opinion is but I am very strongly against this sort of extension model (we have admission, or controllers, but not both together)

mikedanese commented:
Ok, talked to liggit on slack. We are going to explore some other options. I think I misunderstood his comment on the original issue.

pipejakob commented:
I'll update the PR description and release notes, but I've pushed a new commit that:

  - Adds a new command for running just the certificates controller stand-alone
    - Copies the `OWNERS` file from `kube-controller-manager` to bootstrap the new command's ownership (suggestions welcome)
  - Removes the `kube-controller-manager` integration with the new webhook signer, instead only adding support to the stand-alone `certificates-controller`
  - Removes the custom configuration file for the webhook signer, instead taking only the kubeconfig and retry backoff via CLI options

Care to take another look, @liggitt?

mikedanese commented on cmd/certificates-controller/OWNERS:
```diff
@@ -0,0 +1,67 @@
+approvers:
```
where'd this come from?

pipejakob replied:
Copied verbatim from `cmd/kube-controller-manager/OWNERS`, since the code and responsibility has a lot in common with `kube-controller-manager`. I could also bootstrap it from `pkg/controller/certificates/OWNERS` instead, or omit it entirely and fall back on `cmd/OWNERS`. I'm definitely open to suggestions on the best route to bootstrap this.

roberthbailey replied:
With the pending rename, please just put you, me, and mike as owners/approvers of this directory. Other folks can send PRs if they want to be added to the list.

pipejakob replied:
Sounds good.

bgrant0607 commented:
Where is the proposal for this feature?

I agree with @smarterclayton and @liggitt that a webhook-based controller is not compatible with the design of Kubernetes.

Kubernetes provides a collection of similar REST APIs supporting primarily CRUD operations on (mostly) persistent resources.   

Both human-centric clients and components acting as asynchronous controllers interact with the same API resources and shared state. Most resources contain metadata, including labels and annotations, fully elaborated desired state (spec), including default values, and observed state (status). Controllers work continuously to drive the actual state towards the desired state, while reporting back the currently observed state for users and for other controllers to consume. 

The built-in asynchronous controllers are currently run by the Controller Manager, but the intent is that additional controller components could be added as needed.

I intend to ensure that there is a standard, built-in add-on manager in order to facilitate the addition of new cluster components.

mikedanese commented:
@bgrant0607 this controller uses a webhook in the same way that the service controller uses a webhook to create a load balancer. This PR introduces a new controller process that will run on GKE masters like the ingress controller runs on GKE masters. We can't run this as an addon because we need specific auth scopes that will not be available on the nodes. I don't know how you are recommending that this PR should change.

liggitt commented:
> @bgrant0607 this controller uses a webhook in the same way that the service controller uses a webhook to create a load balancer. This PR introduces a new controller process that will run on GKE masters like the ingress controller runs on GKE masters. We can't run this as an addon because we need specific auth scopes that will not be available on the nodes. I don't know how you are recommending that this PR should change.

The ingress controller satisfies the intent expressed in the Kube API by making calls to GCE load balancer APIs, and recording the results. It doesn't talk to the Kube API on one side and a webhook that looks like a Kube API on the other. A webhook controller speaking the Kube API in two directions seems like one too many layers of abstraction to me.

mikedanese commented:
So if we coded this to the specifics of the GKE API it would be more acceptable?

liggitt commented:
yeah, I think the ingress controller is a good example. a standalone controller that uses the GKE API to satisfy CSR requests makes sense to me

pipejakob commented:
@roberthbailey, @mikedanese, @liggitt and I just talked out of band, and here's the summary:

There were concerns around the support and longevity of this code, given that the certificates API is only in beta, and different customers trying to integrate with different external CAs might have different API requirements that aren't satisfied by the webhook API (so it wouldn't save them much work). Since this design was specifically going to be used for GKE, and we modeled our own API after the webhook, we should thoroughly brand it to be GKE-specific. Anyone else could still use it if they find it useful, but it lessens the support burden by making it clear that it's not generically reusable, and gives us the flexibility to change it (and our own API) in the future.

So, the agreed upon plan of action is:

  - Keep the code in-tree for now, but nested under a single `cmd` directory so it's easy to pull it out in the future.
  - Add a TODO to migrate to its own repository in the future.
  - Rename the command to be prefixed with `gke-` and any references from `webhook` to `gke-endpoint` or similar.

@bgrant0607 What do you think about this solution?

pipejakob committed:
New command: gke-certificates-controller

This adds a new stand-alone certificates controller for use on GKE. It
allows calling GKE to sign certificates instead of requiring the CA
private key locally.

It does not aim for 100% feature parity with kube-controller-manager
yet, so for instance, leader election support is omitted.

bgrant0607 commented:
Sounds fine, thanks. I want to be careful about expanding the exposed surface area with cloud providers

roberthbailey commented:
Once we have cloud controller managers, this should move into the gke controller manager binary (which should be outside of the main repo).

pipejakob commented:
Okay, I believe I've adequately rebranded this new command as GKE-specific, and the only references to webhook left are implementation details around the use of `GenericWebhook`. PTAL.

liggitt commented on build/common.sh:
```diff
@@ -102,6 +103,7 @@ kube::build::get_docker_wrapped_binaries() {
           kube-scheduler,armel/busybox
           kube-aggregator,armel/busybox
           kube-proxy,gcr.io/google_containers/debian-iptables-arm:${debian_iptables_version}
+          gke-certificates-controller,busybox
```
don't the rest of these need arch specific prefixes of busybox?

pipejakob replied:
Good catch of bad copy/paste (I had only tested amd64)! Updated the other references.

liggitt commented on cmd/gke-certificates-controller/app/gke_certificates_controller.go:
```diff
@@ -0,0 +1,114 @@
+// getClientBuilder returns the controller.ControllerClientBuilder to use based
+// on the configuration.
+func (s *GKECertificatesController) getClientBuilder(kubeconfig *restclient.Config, kubeClient *clientset.Clientset) controller.ControllerClientBuilder {
+	if len(s.ServiceAccountKeyFile) > 0 && s.UseServiceAccountCredentials {
```
for a controller we expect to run in a pod (I'm assuming that's what we plan to do with this), I'd just take a kubeconfig and/or use in-cluster-config

liggitt commented on cmd/gke-certificates-controller/app/options.go:
```diff
@@ -0,0 +1,69 @@
+// GKECertificatesController is the main context object for the package.
+type GKECertificatesController struct {
+	Kubeconfig                    string
+	Master                        string
```
any reason not to just use the kubeconfig?

pipejakob replied:
Even though I wasn't shooting for full parity with `kube-controller-manager`, I was attempting to make the relevant parts of the CLI match to make the transition easier. Given your other comments, I can certainly ditch that approach and pare down even more here to the bare essentials.

liggitt replied:
Every flag and option is command surface area to support. +1 for bare essentials

liggitt commented on cmd/gke-certificates-controller/app/options.go:
```diff
@@ -0,0 +1,69 @@
+	ClusterSigningGKERetryBackoff metav1.Duration
+	ApproveAllKubeletCSRsForGroup string
+	UseServiceAccountCredentials  bool
+	ServiceAccountKeyFile         string
```
don't need UseServiceAccountCredentials and ServiceAccountKeyFile... that was only needed for the main kube-controller-manager because it was starting a dozen controller loops... it's reasonable to require the kubeconfig (or in-cluster config) to be the credentials this single controller loop will run with

liggitt commented on cmd/gke-certificates-controller/app/options.go:
```diff
@@ -0,0 +1,69 @@
+	fs.StringVar(&s.ClusterSigningGKEKubeconfig, "cluster-signing-gke-kubeconfig", s.ClusterSigningGKEKubeconfig, "If set, use the kubeconfig file to call GKE to sign cluster-scoped certificates instead of using a local private key.")
+	fs.DurationVar(&s.ClusterSigningGKERetryBackoff.Duration, "cluster-signing-gke-retry-backoff", s.ClusterSigningGKERetryBackoff.Duration, "The initial backoff to use when retrying requests to GKE. Additional attempts will use exponential backoff.")
+
+	fs.StringVar(&s.ApproveAllKubeletCSRsForGroup, "insecure-experimental-approve-all-kubelet-csrs-for-group", s.ApproveAllKubeletCSRsForGroup, "The group for which the controller-manager will auto approve all CSRs for kubelet client certificates.")
```
I plan to switch this to a SubjectAccessReview check instead of a hard-coded group. Will it be a problem to deprecate/remove this along with the controller-manager option?

pipejakob replied:
I'll defer to @mikedanese here because I believe he was set on using the GroupApprover for GKE in 1.6 and I don't know how much of a wrench SubjectAccessReview would throw into those plans.

liggitt replied:
Given the timeframe, keeping this for 1.6 is reasonable. Just unfortunate to add options I'd hoped to deprecate shortly

liggitt commented on cmd/gke-certificates-controller/app/options.go:
```diff
@@ -0,0 +1,69 @@
+func NewGKECertificatesController() *GKECertificatesController {
+	s := &GKECertificatesController{
+		ClusterSigningCertFile:        "/etc/kubernetes/ca/ca.pem",
+		ClusterSigningKeyFile:         "/etc/kubernetes/ca/ca.key",
```
I wouldn't default these... default paths that tolerate files not existing make commands hard to reason about (see behavior of kubelet with --kubeconfig and --require-kubeconfig)

liggitt replied:
without defaults, if the paths are set, you can validate they exist and error if they don't. you can also error if paths AND a gke kubeconfig are both set.

pipejakob replied:
Definitely fair. This also falls under the "cloning kube-controller-manager's flags" umbrella. With the GKE-specific rebranding, I believe I can tear these out entirely since we'll always want to call the GKE endpoint instead of having the CA private keys on the masters.

pipejakob commented:
Okay, I've killed service account support and the master flag, so we're just left with requiring a kubeconfig file. Since we'll be running this as a static pod, I don't think we can use InClusterConfig.

liggitt commented:
> Since we'll be running this as a static pod, I don't think we can use InClusterConfig.

seems reasonable as a starting point.

pipejakob commented:
(Rebased in order to fix verify failure)

pipejakob commented:
I've proactively rebased again to make sure this wasn't broken by upstream commits.

I have more changes after this, but I've broken them apart into two follow-on PRs to keep from too much scope creep here (https://github.com/kubernetes/kubernetes/pull/42011 and https://github.com/kubernetes/kubernetes/pull/42030). Will you have time to take a look before the weekend, @liggitt? I'd love to beat the code freeze if I can. Thanks!

ixdy commented:
I don't fully understand what this is doing, but I'm a little worried about adding something that appears to be GKE-specific to the kubernetes-server tarball. Is there a reason we need to sideload this instead of treating it more like an addon and pulling from a container registry?

liggitt commented:
> Will you have time to take a look before the weekend, @liggitt? I'd love to beat the code freeze if I can.

At a glance it looks like my comments were addressed. @mikedanese can have final lgtm on the controller. 

> I don't fully understand what this is doing, but I'm a little worried about adding something that appears to be GKE-specific to the kubernetes-server tarball.

The code is intended to be moved out of tree asap. I agree that the packaging is odd. Is there any way to just let GKE package this?

roberthbailey commented:
I don't see any reason we can't package it separately as long as the code can live in the main repo for a short while longer.

pipejakob commented:
@ixdy My original intent was to reuse this mechanism because it lets us thoroughly e2e test new changes (even against GKE clusters) in order to prevent regressions before they happen. But given these comments, I agree that the right thing to do is remove the packaging from OSS and have it live outside of this repo.

I'll update the PR again to only include the code and Makefile changes necessary to build the single static binary, and worry about building the image (and other cluster configuration behavior) elsewhere. This should also make my other two chained PRs superfluous.

pipejakob commented:
Done. My updated commit is just the `gke-certificates-controller` code and changes necessary to build the binary and pass verification.

pipejakob commented:
@k8s-bot non-cri e2e test this
@k8s-bot kops aws e2e test this

spxtr commented:
/approve

ixdy commented on hack/lib/golang.sh:
```diff
@@ -24,6 +24,7 @@ readonly KUBE_GOPATH="${KUBE_OUTPUT}/go"
 # If you update this list, please also update build/release-tars/BUILD.
 kube::golang::server_targets() {
   local targets=(
+    cmd/gke-certificates-controller
```
this still ends up in `kubernetes-server.tar.gz`. :( but at least there's not a docker tarfile, too.

pipejakob replied:
Ah. Is it more appropriate to add the target to `KUBE_ALL_TARGETS` instead? I don't actually need it in the tar, I just want to make sure that it gets built by default when anyone runs `make` so we'll prevent anyone from breaking the code.

ixdy replied:
that would probably work! I'm not sure whether we want to make another grouping for this (I don't know what to call it) or just add to the list directly.

ixdy commented on hack/lib/golang.sh:
```diff
@@ -24,6 +24,7 @@ readonly KUBE_GOPATH="${KUBE_OUTPUT}/go"
 # If you update this list, please also update build/release-tars/BUILD.
```
nobody reads my comment. :(

pipejakob replied:
To be completely honest, I read the comment immediately before it that said:

    # Note: if you are adding something here, you might need to add it to
    # kube::build::source_targets in build/common.sh as well.

and must have stopped reading because `source_targets` doesn't exist in `build/common.sh` (or anywhere else in the codebase, says `grep -r`). :-)

Given that we don't actually want this in the tar, and I just want to ensure that the binary gets built when users run `make` (or `make bazel-build`), I'll just leave this alone and move the target to `KUBE_ALL_TARGETS`.

ixdy replied:
that's only been gone for 4.5 months (#30787). :)

mind removing that part of the comment while you're here?

pipejakob replied:
Done and done. I could create a new target group like `KUBE_EXTRA_TARGETS` just for this binary, but it seemed overkill for something that's meant to be temporary. Plus, the fact that this target sticks out like a sore thumb in `KUBE_ALL_TARGETS` is a good reminder that we should remove it.

ixdy commented:
/lgtm

Thanks!

pipejakob commented:
@k8s-bot unit test this

pipejakob commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this
